### PR TITLE
.Net: Function Calling Planner: Catch exceptions and output error messages to chat history

### LINKED ISSUE #3224: .Net: How to output error messages to [OBSERVATION] in native functions
My native function raises an exception with a message when an error occurs.
`throw new Exception( "Argument must not be a negative number" );`

However, this error message is not output to [OBSERVATION].
In the log, the exception message always looks like this.
`2023-xx-xx xx:xx:xx.xxxx|INFO|Microsoft.SemanticKernel.Planners.StepwisePlanner|Observation: Error invoking action MyPlugin.Sqrt : Exception has been thrown by the target of an invocation.`

I need a way to output error messages to [OBSERVATION].

### PR DESCRIPTION
**Pull request by gitri-ms**
### Motivation and Context

 Resolves #3224. Exceptions thrown by invoked functions are currently not being caught by the planner.

### Description

This change catches any non-critical exceptions from function invocation, adds the error message into chat history, and lets the planner retry.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Review by glahaye — APPROVED** [review]

File: dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs [hunk 1]
```diff
@@ -139,6 +140,10 @@ public sealed class FunctionCallingStepwisePlanner
                     {
                         chatHistoryForSteps.AddUserMessage($"Failed to execute function {functionResponse.FullyQualifiedName}. Try something else!");
                     }
+                    catch (Exception ex) when (!ex.IsCriticalException())
```
Comment (alliscode) [inline dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs:143]: Can we add a test case for this functionality?

File: dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs [hunk 2]
```diff
@@ -139,6 +140,10 @@ public sealed class FunctionCallingStepwisePlanner
...
+                    {
+                        chatHistoryForSteps.AddUserMessage($"Execution of function {functionResponse.FullyQualifiedName} failed with error: \"{ex.Message}\" Try something else!");
```
Comment (gitri-ms) [inline dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs:145]: Need to add tool message here, per this error: `An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'.`

File: dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs (hunk 1 above)
Comment (gitri-ms) [reply dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs:143]: Working on it!

File: dotnet/src/IntegrationTests/Planners/Stepwise/FunctionCallingStepwisePlannerTests.cs [hunk 3]
```diff
@@ -72,6 +72,28 @@ public async Task CanExecuteStepwisePlanAsync(string prompt, string[] expectedFu
         }
     }
 
+    [Fact]
+    public async Task DoesNotThrowWhenPluginFunctionThrowsAsync()
```
Comment (alliscode) [inline dotnet/src/IntegrationTests/Planners/Stepwise/FunctionCallingStepwisePlannerTests.cs:76]: nit: DoesNotThrowWhenPluginFunctionThrowsNonCriticalExceptionAsync

Comment (alliscode) [reply dotnet/src/IntegrationTests/Planners/Stepwise/FunctionCallingStepwisePlannerTests.cs:76]: Could you test the critical exception case as well.

**Review by glahaye — APPROVED** [review]

**Review by alliscode — APPROVED** [review]

**Pull request merged**